stdlib/experimental/diff.go

```diff
@@ -0,0 +1,717 @@
+	if err != nil {
+		return table.Chunk{}, false, err
+	}
+
+	// Check if this was a perfect match. The lengths have to be the same.
+	if want.Len() == got.Len() {
+		isPerfectMatch := true
+		for i, n := 0, want.Len(); i < n; i++ {
+			if !schema.equal(d, i, i) {
+				isPerfectMatch = false
+				break
+			}
+		}
+
+		if isPerfectMatch {
+			return table.Chunk{}, false, nil
+		}
+	}
+
+	// At this point, we are producing a diff in some manner and
+	// we expect to find some difference.
+	diff := array.NewStringBuilder(d.mem)
+	builders := make([]array.Builder, len(schema.wantIdx))
+	for i, col := range schema.cols[schema.offset:] {
+		builders[i] = arrow.NewBuilder(col.Type, d.mem)
+	}
+
+	// Compute the lcs (longest common subsequence) table.
+	lcs := d.lcs(schema)
+	tracer := diffTrace{
+		lcs:      lcs,
+		schema:   schema,
+		diff:     diff,
+		builders: builders,
+	}
+	// Trace the table from the last element.
+	tracer.trace(len(lcs[0])-1, len(lcs)-1)
+
+	// Create the diff table.
+	buf := arrow.TableBuffer{
+		GroupKey: key,
+		Columns:  schema.cols,
+		Values:   make([]array.Array, 0, len(schema.cols)),
+	}
+	buf.Values = append(buf.Values, diff.NewArray())
+	for i, col := range key.Cols() {
+		arr := arrow.Repeat(col.Type, key.Value(i), diff.Len(), d.mem)
+		buf.Values = append(buf.Values, arr)
+	}
+	for _, b := range builders {
+		buf.Values = append(buf.Values, b.NewArray())
+	}
+	return table.ChunkFromBuffer(buf), true, nil
+}
+
+type diffTrace struct {
+	lcs      [][]diffLcsEntry
+	schema   *diffSchema
+	diff     *array.StringBuilder
+	builders []array.Builder
+}
+
+func (d *diffTrace) trace(row, col int) {
+	parent := d.lcs[col][row].parent
+	if row > 0 && col > 0 && parent == diffParentMatch {
+		d.trace(row-1, col-1)
+		d.diff.Append("")
+		d.schema.appendRow(d.builders, 0, row-1)
+	} else if row > 0 && (col == 0 || parent == diffParentLeft) {
+		d.trace(row-1, col)
+		d.diff.Append("-")
+		d.schema.appendRow(d.builders, 0, row-1)
+	} else if col > 0 && (row == 0 || parent == diffParentTop || parent == diffParentEither) {
+		d.trace(row, col-1)
+		d.diff.Append("+")
+		d.schema.appendRow(d.builders, 1, col-1)
+	}
+}
+
+type diffLcsParent int
+
+const (
+	diffParentEither diffLcsParent = iota
+	diffParentLeft
+	diffParentTop
+	diffParentMatch
+)
+
+type diffLcsEntry struct {
+	parent diffLcsParent
+	length int
+}
+
+// lcs computes a longest common subsequence table.
+//
+// This constructs a table like the one present here: https://en.wikipedia.org/wiki/Longest_common_subsequence_problem#Traceback_approach.
+//
+// The table is constructed with the row dimension corresponding to the want
+// and the column dimension corresponding to the got.
+//
+// The table itself is accessed with table[col][row]. The table itself
+// is also 1 indexed instead of zero where zero is the base case. That means the first
+// element in the table chunks corresponds to index 1 in this table.
+//
+// See the example in the wikipedia article for details.
```

Comment:
So is the memory used for this _O(n*m)_ where _n_ and _m_ is the number of rows on the want and got tables?

Reply:
Yes although it's `O(n+1*m+1)` because of the extra row/column for the base case.